test/signalrclienttests/hub_connection_impl_tests.cpp

```diff
@@ -1069,4 +1084,149 @@ TEST(progress, exceptions_from_progress_callbacks_logged)
+                .then([test_completed_event, hub_connection](pplx::task<void> invoke_void_task)
+                {
+                    try
+                    {
+                        invoke_void_task.get();
+                        ASSERT_TRUE(false); // exception expected but not thrown
+                    }
+                    catch (const std::exception& e)
+                    {
+                        ASSERT_STREQ("\"connection has been lost\"", e.what());
+                    }
+                });
+
+            message_sent_event->set();
+
+            return invoke_task;
+        }).get();
+}
+
+TEST(reconnect, pending_invocations_finished_and_custom_reconnecting_callback_invoked_if_connection_lost)
+{
+    auto message_sent_event = std::make_shared<pplx::event>();
+
+    auto init_sent = false;
+    auto websocket_client = create_test_websocket_client(
+        /* receive function */ [init_sent, message_sent_event]() mutable
+        {
+            if (init_sent)
+            {
+                message_sent_event->wait();
+                return pplx::task_from_exception<std::string>(std::runtime_error("connection exception"));
+            }
+
+            init_sent = true;
+            return pplx::task_from_result<std::string>("{ \"C\":\"x\", \"S\":1, \"M\":[] }");
+        },
+        /* send function */ [](const utility::string_t){ return pplx::task_from_result(); },
+        /* connect function */[](const web::uri& url)
+        {
+            if (url.path() == _XPLATSTR("/reconnect"))
+            {
+                return pplx::task_from_exception<void>(std::runtime_error("reconnect rejected"));
+            }
+
+            return pplx::task_from_result();
+        });
+
+    auto hub_connection = create_hub_connection(websocket_client);
+    auto reconnecting_invoked = false;
+    hub_connection->set_reconnecting([&reconnecting_invoked](){ reconnecting_invoked = true; });
+
+    auto test_completed_event = std::make_shared<pplx::event>();
+    hub_connection->start()
+        .then([hub_connection, message_sent_event, test_completed_event]()
+        {
+            auto invoke_task = hub_connection->invoke_void(_XPLATSTR("TestHub"), _XPLATSTR("TestMethod"), json::value::array())
+                .then([test_completed_event, hub_connection](pplx::task<void> invoke_void_task)
+                {
+                    try
+                    {
+                        invoke_void_task.get();
+                        ASSERT_TRUE(false); // exception expected but not thrown
+                    }
+                    catch (const std::exception& e)
+                    {
+                        ASSERT_STREQ("\"connection has been lost\"", e.what());
+                    }
+                });
+
+            message_sent_event->set();
+
+            return invoke_task;
+        }).get();
+
+    ASSERT_TRUE(reconnecting_invoked);

```

Comment:
I had quite a few problems writing this test but have not seen this on my box. Probably need to use an event to synchronize instead of just a bool.